File: app/src/main/java/com/vkochenkov/weatherfromopenapis/presentation/activities/CitiesListActivity.java
package com.vkochenkov.weatherfromopenapis.presentation.activities;

import android.app.Dialog;
import android.content.Intent;
import android.database.sqlite.SQLiteDatabase;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.vkochenkov.weatherfromopenapis.R;
import com.vkochenkov.weatherfromopenapis.data.db.DBHelper;
import com.vkochenkov.weatherfromopenapis.presentation.dialogs.CityAddDialog;
import com.vkochenkov.weatherfromopenapis.presentation.dialogs.CityDeleteDialog;
import com.vkochenkov.weatherfromopenapis.data.City;
import com.vkochenkov.weatherfromopenapis.presentation.recycler.CityClickListener;
import com.vkochenkov.weatherfromopenapis.presentation.recycler.CityListAdapter;

public class CitiesListActivity extends AppCompatActivity {

    private Toolbar toolbar;
    private EditText citiesSearchField;
    private RecyclerView cityListView;
    private Button addCityBtn;

    private DBHelper dbHelper;
    private SQLiteDatabase database;

    @Override
    protected void onResume() {
        super.onResume();
        Log.d("onResumed", "onResumed");
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_cities_list);

        initFields();
        initRecycler();
        setAddCityBtnClickListener();
        setTextWatcherForSearchCity();

        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onBackPressed();
            }
        });
    }

    private void initRecycler() {
        cityListView.setLayoutManager(new LinearLayoutManager(this));
        CityClickListener cityClickListener = new CityClickListener() {
            @Override
            public View.OnClickListener onCityClickListener(final City city) {
                return new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        Intent intent = new Intent();
                        intent.putExtra("latitude", city.getLatitude());
                        intent.putExtra("longitude", city.getLongitude());
                        setResult(RESULT_OK, intent);
                        finish();
                    }
                };
            }

            @Override
            public View.OnLongClickListener onCityLongClickListener(int position, final City city) {

                return new View.OnLongClickListener() {
                    @Override
                    public boolean onLongClick(View view) {

                        Dialog dialog = new CityDeleteDialog(CitiesListActivity.this, dbHelper, cityListView, city);
                        dialog.show();

                        return true;
                    }
                };
            }

        };
        cityListView.setAdapter(new CityListAdapter(database, cityClickListener));

        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(this,
                                                                                this.getResources().getConfiguration().orientation);
        cityListView.addItemDecoration(dividerItemDecoration);
    }

    private void initFields() {
        toolbar = findViewById(R.id.toolbar_cities_list);
        citiesSearchField = findViewById(R.id.edt_city_search_field);
        cityListView = findViewById(R.id.recycler_city_list);
        addCityBtn = findViewById(R.id.add_city_btn);

        dbHelper = new DBHelper(this);
        database = dbHelper.getWritableDatabase();
    }

    private void setAddCityBtnClickListener() {
        addCityBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Dialog dialog = new CityAddDialog(CitiesListActivity.this, dbHelper, cityListView);
                dialog.show();
            }
        });
    }

    private void setTextWatcherForSearchCity() {
        citiesSearchField.addTextChangedListener(new TextWatcher() {

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                CityListAdapter adapter = (CityListAdapter) cityListView.getAdapter();
                if (adapter != null) {
                    adapter.findAndShowCityItems(charSequence.toString());
                }
            }

            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
            }
        });
    }
}

File: app/src/main/java/com/vkochenkov/weatherfromopenapis/presentation/activities/CitiesListActivity.kt
package com.vkochenkov.weatherfromopenapis.presentation.activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.View.OnLongClickListener
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.vkochenkov.weatherfromopenapis.R
import com.vkochenkov.weatherfromopenapis.data.db.entities.City
import com.vkochenkov.weatherfromopenapis.presentation.dialogs.CityAddDialog
import com.vkochenkov.weatherfromopenapis.presentation.dialogs.CityDeleteDialog
import com.vkochenkov.weatherfromopenapis.presentation.recycler.CityClickListener
import com.vkochenkov.weatherfromopenapis.presentation.recycler.CityListAdapter
import com.vkochenkov.weatherfromopenapis.presentation.viewmodel.CitiesListViewModel

class CitiesListActivity : AppCompatActivity() {

    private val viewModel: CitiesListViewModel by lazy {
        ViewModelProvider(this).get(CitiesListViewModel::class.java)
    }

    private lateinit var toolbar: Toolbar
    private lateinit var citiesSearchField: EditText
    private lateinit var cityListView: RecyclerView
    private lateinit var addCityBtn: FloatingActionButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_cities_list)
        initFields()
        initRecycler()
        setAddCityBtnClickListener()
        initObserveForViewModel()

        viewModel.getAllCitiesFromDb()
        setTextWatcherForSearchCity()
        toolbar.setNavigationOnClickListener { onBackPressed() }
    }

    private fun initObserveForViewModel() {
        viewModel.citiesLiveData.observe(this, Observer {
            (cityListView.adapter as CityListAdapter).setData(it)
        })
    }

    private fun initRecycler() {
        cityListView.layoutManager = LinearLayoutManager(this)
        val cityClickListener: CityClickListener = object : CityClickListener {

            override fun onCityClickListener(city: City): View.OnClickListener {
                return View.OnClickListener {
                    val intent = Intent()
                    intent.putExtra("latitude", city.latitude);
                    intent.putExtra("longitude", city.longitude);
                    setResult(Activity.RESULT_OK, intent)
                    Toast.makeText(
                        getApplicationContext(),
                        "${city.name}", Toast.LENGTH_SHORT
                    ).show()
                    finish()
                }
            }

            override fun onCityLongClickListener(position: Int, city: City): OnLongClickListener {
                return OnLongClickListener {
                    var dialog = CityDeleteDialog(this@CitiesListActivity, viewModel, city)
                    dialog.show()
                    true
                }
            }
        }
        cityListView!!.adapter = CityListAdapter(cityClickListener)
        val dividerItemDecoration = DividerItemDecoration(
            this,
            this.resources.configuration.orientation
        )
        cityListView!!.addItemDecoration(dividerItemDecoration)
    }

    private fun initFields() {
        toolbar = findViewById(R.id.toolbar_cities_list)
        citiesSearchField = findViewById(R.id.edt_city_search_field)
        cityListView = findViewById(R.id.recycler_city_list)
        addCityBtn = findViewById(R.id.add_city_btn)
    }

    private fun setAddCityBtnClickListener() {
        addCityBtn.setOnClickListener {
            val dialog = CityAddDialog(this@CitiesListActivity, viewModel)
            dialog.show()
        }
    }

    private fun setTextWatcherForSearchCity() {
        citiesSearchField.addTextChangedListener(object : TextWatcher {
            override fun onTextChanged(
                charSequence: CharSequence,
                i: Int,
                i1: Int,
                i2: Int
            ) {
                viewModel.getSameNameCitiesFromDb(charSequence.toString())
            }

            override fun beforeTextChanged(
                charSequence: CharSequence,
                i: Int,
                i1: Int,
                i2: Int
            ) {
            }

            override fun afterTextChanged(editable: Editable) {}
        })
    }
}